src/components/Header/index.jsx

```diff
@@ -152,88 +151,85 @@ const GordonHeader = ({ onDrawerToggle }) => {
   );
 
   return (
-    <section className={styles.gordon_header}>
-      <AppBar className={styles.app_bar} position="static">
-        <Toolbar>
-          <IconButton
-            className={styles.menu_button}
-            color="primary"
-            aria-label="open drawer"
-            onClick={onDrawerToggle}
-            size="large"
-          >
-            <MenuIcon className={styles.menu_button_icon} />
-          </IconButton>
-
-          <Typography className={`disable_select ${styles.title}`} variant="h6" color="inherit">
-            <Routes>
-              {routes.map((route) => (
-                <Route
-                  key={route.path}
-                  path={route.path}
-                  element={() => {
-                    setDocumentTitle(route.name || projectName);
-                    return <span>{route.name || projectName}</span>;
-                  }}
-                />
-              ))}
-            </Routes>
-          </Typography>
-
-          <div className={styles.center_container}>
-            <Tabs
-              textColor="inherit"
-              indicatorColor="secondary"
-              centered
-              value={tabIndex}
-              onChange={(event, value) => setTabIndex(value)}
+    <>
+      {updateTabHighlight()}
```

Comment:
I have a couple of comments about this.

Firstly, it's a little bit odd to call the function here, inside of the return statement and nested in the JSX Fragment. You could achieve exactly the same effect by calling the function anywhere else in the component after it's defined, so it's unnecessarily confusing to put it here.

Secondly, it's not best practice to define a function and then call it inside the body of a component. That is because, every time the component renders, it will recreate the function and then call it again. That is causing the component to do a lot of unnecessary work. Also, notice that you're calling `setTabIndex` inside the function, which means that every time this component renders, if the tab index should change, you're immediately changing the state, which will cause another re-render.

More specifically, when `location` changes (because the user navigated to a new page), `useLocation` will update - which counts as a state update the same way that calling `setState` causes a state update. That will cause the Header component to re-render with the new value of `location`. Then, while the component is re-rendering, it will re-evaluate the definition of `updateTabHighlight` and re-create the function. Finally, the component will reach the `return` statement and begin interpreting the JSX Fragment - at which point it reaches this line and runs `updateTabHighlight`. Because `location` has changed, there's a good chance that `tab !== tabIndex` will be true, so we will call `setTabIndex`, which will cause the component to re-render again.

All that to say - we're doing more work than is necessary in a round-about way here. Because `tabIndex` can be calculated based only on `location`, it doesn't need to be it's own state variable, it can just be a regular javascript variable.

There a few different ways to clean this up, but I think the best way is actually by creating a custom hook, which would look something like this:
```javascript

const TabUrlPatterns = [
      /^\/$/,
      /^\/involvements\/?$|^\/activity/,
      /^\/events\/?$/,
      /^\/people$|^\/myprofile|^\/profile/,
      /^\/timesheets$/,
    ];

const useTabHighlight = () => {
  const location = useLocation();
  const [tabIndex, setTabIndex] = useState(0);

  useEffect(() => {
    const matchedIndex = TabUrlPatterns.findIndex((pattern) => pattern.test(location));
    setTabIndex(matchedIndex); // This won't cause an update if the new value is the same as the old value
  }, [location]);

  return tabIndex;
};
```

Then you can replace all the tab highlight code in Header with this:
```javascript
  const tabIndex = useTabHighlight();
```

Reply:
That makes sense! Thank you for writing this whole paragraph for me! I did the changes you suggested, and it works perfectly. There is only one thing that I don't understand. If I use `location` here in the code `const matchedIndex = TabUrlPatterns.findIndex((pattern) => pattern.test(location));`, the highlight is not working at all, which means there is no highlight on any of the tabs no matter which tab I press. However, if I change the `location` to `location.pathname`, then `const matchedIndex = TabUrlPatterns.findIndex((pattern) => pattern.test(location.pathname));` works very well. I don't understand why simply adding a `.pathname` solve the problem, but I believe basically the problem is solved.

Reply:
Sorry, that was my bad using `location`. The `location` returned by `useLocation` is an object like this:
```typescript
interface Location {
  pathname: string;
  search: string;
  hash: string;
  state: unknown;
  key: string;
}
```
So naturally, that object won't match any of the Regular Expressions that are searching for a string.

Reply:
I see! Thanks!